Incremental directory reading

patricoferris opened the pull request:
When first providing `readdir` we shelved an incremental version. This PR starts the ball rolling on adding support for something like [`with_dir_entries`](https://github.com/ocaml-multicore/eio/pull/207#pullrequestreview-987295774). Opening early to get some feedback on how it should look, should `Eio.Path.walk` walk directory tree recursively or just a single directory, leaving it to the user to descend should they want to. We should perhaps have a `type entry` too.

I wrote a small benchmark:

```ocaml
open Eio.Std

let ( / ) = Eio.Path.( / )

let dirname i = Fmt.str "dir-%i" i
let filename i = Fmt.str "file-%i" i

let make_bench_dir ?(depth=1) ?(width=1000) dir =
  let rec loop dir depth =
    if depth <= 0 then ()
    else (
      Eio.Path.mkdir ~perm:0o700 (dir / dirname depth);
      Fiber.List.iter ~max_fibers:10 (fun fname ->
        Eio.Path.save ~create:(`If_missing 0o644) (dir / fname) "hello world"
      ) (List.init width filename);
      loop (dir / dirname depth) (depth - 1)
    )
  in
  loop dir depth

let read_dir_walk ~f dir =
  let rec aux acc path =
    match Eio.Path.kind ~follow:true path with
    | `Directory as k ->
       let acc2 = f acc (k, path) in
       let files = Eio.Path.read_dir path in
       List.fold_left (fun acc f -> aux acc (path / f)) acc2 files
    | k -> f acc (k, path)
    in
  let files = Eio.Path.read_dir dir in
  List.fold_left (fun acc f -> aux acc (dir / f)) [] files

let rec walk ~f acc dir =
  Eio.Path.walk dir @@ fun seq ->
  Seq.fold_left (fun acc -> function
    | (`Directory, path) ->
      let acc2 = f acc (`Directory, dir / path) in
      walk ~f acc2 (dir / path)
    | (kind, fname) -> f acc (kind, dir / fname)) acc seq

let time fn =
  let now = Mtime_clock.now_ns () in
  let v = fn () in
  let stop = Mtime_clock.now_ns () in 
  Eio.traceln "Took: %a" Fmt.uint64_ns_span (Int64.sub stop now);
  v

module Path = struct
  include Eio.Path
  type t = Eio.Fs.dir_ty Eio.Path.t 

  let compare a b = String.compare (Eio.Path.native_exn a) (Eio.Path.native_exn b)
end

module S = Set.Make (Path)

let () = 
  Eio_main.run @@ fun env ->
  let tmpdirf = Filename.temp_dir "eio-walk-" "-bench" in
  let tmpdir = env#fs / tmpdirf in
  make_bench_dir ~depth:15 ~width:20_000 tmpdir;
  let f acc (_, path) = path :: acc in
  let files2 = time (fun () -> walk ~f [] tmpdir) |> S.of_list in
  let files1 = time (fun () -> read_dir_walk ~f tmpdir) |> S.of_list in
  Eio.traceln "Files 1: %i\n" (S.cardinal files1);
  Eio.traceln "Files 2: %i\n" (S.cardinal files2);
  Eio.traceln "Same files? %b\n" (S.equal files1 files2);
  let _ : int = Sys.command (Fmt.str "rm -r %s" tmpdirf) in
  ()
```

It is far from perfect and we should take into account FS caching, but on my machine I'm getting the following.

For Eio_linux: 1.46s to 110ms
For Eio_posix: 2.02s to 1.58s

patricoferris committed:
Add with_dir_entries for walking a filesystem

patricoferris commented:
Thanks @avsm !

Note that the current implementation does _not_ descend into directories (see the recursive call to `walk` in the benchmark above). Happy to change the behaviour to recurse by default.

avsm commented:
Got it! Your approach makes more sense to me. We can build a recursive descender _over_ this interface in the future.

avsm commented:
Just chatting to @patricoferris, I think this version that _doesnt_ recurse is best, since we can then add a parallel or sequential recursing function over this.

talex5 commented on lib_eio/unix/eio_unix_stubs.c:
```diff
@@ -3,11 +3,15 @@
 #include <caml/bigarray.h>
+#include <caml/alloc.h>
+
+#define BUF_SIZE 4096
```
This doesn't seem to be used.

talex5 commented on lib_eio/path.ml:
```diff
@@ -106,6 +106,14 @@ let read_dir t =
+  try X.with_dir_entries dir path fn
+  with Exn.Io _ as ex ->
+    let bt = Printexc.get_raw_backtrace () in
+    Exn.reraise_with_context ex bt "walking directory %a" pp t
```
Not sure adding the exception handler here is a good idea. This is likely to be called recursively, and we don't want all the parents adding their own error context. Probably the backends should wrap just the actual call.

talex5 commented on lib_eio/path.mli:
```diff
@@ -155,6 +155,9 @@ val read_dir : _ t -> string list
     Note: The special Unix entries "." and ".." are not included in the results. *)
 
+val walk : _ t -> ((File.Stat.kind * string) Seq.t -> 'a) -> 'a
+(** [walk t] traverses the directory [t] producing a sequence of results. *)
```
Comment needs revising (it doesn't necessarily produce a sequence of results). Should say something about ordering here too.

talex5 commented on lib_eio_linux/low_level.ml:
```diff
@@ -513,11 +513,17 @@ let read_dir fd =
+  Fd.use_exn "read_some_dir" fd @@ fun fd ->
+  Eio_unix.run_in_systhread ~label:"read_some_dir" @@ fun () ->
+  let files = eio_getdents fd in
+  List.filter_map (function _, ".." | _, "." -> None | v -> Some v) files
```
```suggestion
  List.filter (function _, ".." | _, "." -> false | _ -> true) files
```
Maybe we should just filter these out in the C code, though?

talex5 commented on lib_eio_posix/fs.ml:
```diff
@@ -86,6 +86,9 @@ end = struct
     |> Array.to_list
 
+  let with_dir_entries t path fn =
+    Err.run (Low_level.with_dir_entries t.fd path) fn
```
Probably don't want to wrap `fn` as well here, just the actual read.

talex5 commented on lib_eio_posix/low_level.ml:
```diff
@@ -389,6 +390,44 @@ let readdir dirfd path =
     Fd.use_exn "readdir" dirfd @@ fun dirfd ->
     use_confined (Some dirfd)
 
+let with_dir_entries dirfd path fn =
```
Most of this is duplicated from `readdir`. Could perhaps factor it out as `with_dir_handle` or similar?

talex5 commented on lib_eio_posix/low_level.ml:
```diff
@@ -389,6 +390,44 @@ let readdir dirfd path =
+    match eio_readdir h with _, "." | _, ".." -> read_entry h | v -> v
+  in
+  let it h () =
+    match in_worker_thread "with_dir_entries" @@ fun () -> read_entry h with
```
Seems a shame to use a thread job for every entry individually. I'd guess that reading 1000 at a time might speed things up a fair bit. The only problem I see is if you had a really slow network file-system and wanted to get the first entries without waiting.

talex5 replied:
On a test I did (recursively scanning `/usr`), using a batch size of 100 reduced the total time from 3.14s to 1.04s. Increasing the batch size to 1000 didn't give any further improvement.

talex5 commented on lib_eio_posix/low_level.ml:
```diff
@@ -389,6 +390,44 @@ let readdir dirfd path =
+  let use h =
+    let seq = Seq.of_dispenser (it h) in
+    let v = fn seq in
+    Unix.closedir h;
```
If `fn` raises, `h` doesn't get closed. Would probably be simpler to do all the closing in one place.

talex5 reviewed (commented):
Thanks - this looks really good!

I agree we should start with this non-recursive version. Might be good to rename it though (Python uses `walk` for the recursive version: https://docs.python.org/3/library/os.html#os.walk).

(also, you need to install `lintcstubs_arity` to get `primitives.h` to update)

talex5 committed:
Cleanups and suggestions

- Don't wrap `walk` with an exception handler, as that would include the
  user's callback.
- Documentation for `walk`.
- Add a simple test for it.
- Rename it as `with_dir_entries`.
- Add `read_dir_entries` helper that still does sorting and error reporting.

eio_linux:
- Filter out `.` and `..` earlier.

eio_posix:
- Handle errors from `readdir`.
- Open directory in a worker thread too.
- Share opening logic.
- Ensure handle is always closed.
- Don't wrap user's function in our error handling.
- Batch reading directory entries (much faster).

eio_windows:
- Add (very unoptimised) implementation.
- Fix eio/unix stubs build on Windows.

talex5 commented:
I've rebased and updated this as follows:

- Don't wrap `walk` with an exception handler, as that would include the user's callback.
- Improve documentation for `walk`.
- Add a simple test for it.
- Rename it as `with_dir_entries`.
- Add `read_dir_entries` helper that still does sorting and error reporting.

eio_linux:
- Filter out `.` and `..` earlier.

eio_posix:
- Handle errors from `readdir`.
- Open directory in a worker thread too.
- Share opening logic.
- Ensure handle is always closed.
- Don't wrap user's function in our error handling.
- Batch reading directory entries (much faster).

eio_windows:
- Add (very unoptimised) implementation.
- Fix eio/unix stubs build on Windows.

@patricoferris if you're happy with this, I think it's ready to squash and merge once CI is done.

avsm commented on lib_eio/unix/eio_unix_stubs.c:
```diff
@@ -52,3 +54,21 @@ CAMLprim value eio_unix_readlinkat(value v_fd, value v_path, value v_cs) {
+  return caml_hash_variant("Unknown");
+  #else
+  switch (d_type) {
+    case DT_REG: return caml_hash_variant("Regular_file");
```
caml_hash_variant involves computation, so I think this'll happen every time the function is called. Is it worth assigning these to static variables so this conversion is faster? https://github.com/ocaml/ocaml/blob/b5ec56d7afe77b4e57df4ce369c1281a92c6303d/runtime/hash.c#L313

talex5 replied:
I don't mind, but I tried it in godbolt and gcc and clang both evaluated it at compile time:

https://godbolt.org/z/hcMnnKzo4

avsm replied:
Wow, that's impressive (and good enough for me!)

talex5 replied:
Oh, except it doesn't work in the actual code (because `caml_hash_variant` is defined in a .c file, not a header, I think). I'll expand it.

talex5 replied:
It's actually not obvious how to do this. C doesn't seem to have a way to force compile-time evaluation, even though real compilers do it anyway, so it won't let you use `caml_hash_variant` in a static initialiser.

We could get `discover.ml` to write a header file with the values, but getting the value from OCaml isn't easy.
In https://github.com/ocaml/ocaml/issues/4751, the compiler devs say:

> You can already do it for fixed strings by (Obj.magic `Variant : int).
> You can also copy the corresponding function from ocaml sources
> (it is supposed to be fixed forever).

Seems messy, though.

(also, we now seem to have three functions like this: in eio/unix, eio_posix and uring)

avsm commented:
Read rest of diff, lgtm

talex5 commented:
Since the `caml_hash_variant` problem was also there before this PR, let's merge this as it is and optimise the hashing separately. It's strictly an improvement on the previous system of calling `kind` separately for each entry.

talex5 commented:
I tried pre-calculating the hashes here: https://github.com/talex5/eio/commit/575f0b923ab62378eee4030a4da8318595bef22e

However, it didn't make it noticeably faster, so probably not worth it.
